src/pages/AssignServerRoles.js

```diff
@@ -978,6 +999,63 @@ class AssignServerRoles extends BaseWizardPage {
     }
   }
 
+  /**
+   * When a server is edited (which is only possible in from the assigned servers page
+   * on the right), update the details of the matching entry in the list that backs
+   * the available servers (on the left).  This ensures that the correct information
+   * is persisted to the discovered server store, and also ensures that no information
+   * is dropped when unassigning and reassigning servers to roles.
+   */
+  updateServerForEditServer = (server) => {
+    for (let list of ['rawDiscoveredServers', 'serversAddedManually']) {
+      let idx = this.state[list].findIndex(s => server.id === s.id);
+      if (idx >= 0) {
+        let updated_server;
+        this.setState(prev => {
+          let tempList = prev[list].slice();
+          updated_server = this.getMergedServer(tempList[idx], server);
+          tempList.splice(idx, 1, updated_server);
+          return {[list]: tempList};
+        }, () => {
+          this.updateDiscoveredServer(updated_server)
+            .then((response) => {})
+            .catch((error) => {
+              let msg = translate('server.discover.update.error', updated_server.name);
+              this.setState(prev => { return {
+                messages: prev.messages.concat([{msg: [msg, error.toString()]}])
+              };});
+            });
+        });
+        break;
+      }
+    }
+  }
+
+  updateModelObjectForEditServer = (server) => {
+    //update model
+    let model = this.props.model;
+
+    let index = model.getIn(['inputModel', 'servers']).findIndex(e => e.get('id') === server.id);
```

Comment:
This logic is checking for the presence of the server in the model, and if it isn't there, then it will add it. How can this happen? The only way to edit the server is after it has been assigned a role, at which time it has been placed in the model. Removing that would make the logic simpler, something like:
```
let model = this.props.model.updateIn(['inputModel','servers'], list => list.map(svr => {
  if (svr.get('id') === server.id)
    return getMergedServer(svr, server);
  else
    return svr;
}))
```

Reply:
I think that function you refactored...I just restored what was there before....let take a look.

Reply:
I see what you mean. Yeah...it could be simplified. I will make the change and test it.

Reply:
@GarySmith I tried the way you suggested and tested it ...but it doesn't seem to work correctly.  At first I thought I will remove the adding part, but I think I will keep the existing logic for now. Let me know what's your take. Thanks.

Reply:
Why is the adding part needed? Is there some other logic flaw that this is hiding? And why isn't getMergedServer used here like it is in other places where these exact same values are being updated in the model?

Reply:
I can remove the adding logic.